### PR TITLE
podman: network backend control/switch added

### PR DESCRIPTION
**Pull request by m-dati**
Last update for SLEM PC aggregates: added netavark as default also for SLEM5.5 in pcloud, and skipped cni tests.

- Related ticket: https://progress.opensuse.org/issues/137168
- Verification run: See PR body.

### CODE REVIEW

File: tests/containers/podman_network_cni.pm [hunk 1]
```diff
@@ -12,19 +12,21 @@ use testapi;
...
 sub run() {
 
     my ($self, $args) = @_;
     select_serial_terminal;
     my $podman = $self->containers_factory('podman');
 
-
     my $podman_version = get_podman_version();
     my $supports_network = (package_version_cmp($podman_version, '3.1.0') >= 0) ? 0 : 1;
+    # cni removed since podman 5.x
```
Comment (ricardobranco777) [inline tests/containers/podman_network_cni.pm:25]: CNI will be deprecated but not removed on podman 5.x

File: lib/containers/podman.pm [hunk 2]
```diff
@@ -38,4 +40,46 @@ sub get_storage_driver {
...
+    if ($net =~ /^cni/) {
+        @pkgs = qw(cni cni-plugins);
+    } elsif ($net =~ /^netavark/) {
+        @pkgs = qw(netavark aardvark-dns);
+    } else {
+        return 0;
+    }
+    my $config = "/etc/containers/containers.conf";
+    # skip if already installed
+    unless (script_run("rpm -q @pkgs") eq 0) {
```
Comment (ricardobranco777) [inline lib/containers/podman.pm:59]: We never check if packages are installed before installing.

File: lib/containers/podman.pm [hunk 3]
```diff
@@ -38,4 +40,46 @@ sub get_storage_driver {
     return $storage;
 }
 
+sub switch_network_backend {
+    my ($self, $net) = @_;
```
Comment (ricardobranco777) [inline lib/containers/podman.pm:44]: Let's use $backend instead of $net as the latter may be anything from network, etc.

**Review by ricardobranco777 — CHANGES_REQUESTED** [review]

File: lib/containers/podman.pm (hunk 2 above)
Comment (m-dati) [reply lib/containers/podman.pm:59]: Eh in fact: if installed already, it is unuseful to run i.e. transactional update

**Comment by mloviska** [comment]
I am not that confident that we need this as whenever the migration to netavark occurs it has to be done manually. So CNI can be tested before netavark module. Once that CNI is gone, this function wont be useful hence putting it to podman library means that at some point we will need to do a clean up

**Comment by mloviska** [comment]
and for distros that uses `netavark` as default, we will never switch back to CNI

**Comment by m-dati** [comment]
> and for distros that uses `netavark` as default, we will never switch back to CNI
Until we have `podma_network_cni` somewhere running, it could happen, IMHO

File: lib/containers/podman.pm (hunk 2 above)
Comment (mloviska) [reply lib/containers/podman.pm:59]: sure, but in that case you just need if `rpm` returns non-zero

hence `if (script_run("rpm -q @pkgs"))` or `if (script_run("rpm -q @pkgs") != 0)`

**Comment by mloviska** [comment]
> > and for distros that uses `netavark` as default, we will never switch back to CNI
> 
> Until we have `podma_network_cni` somewhere running, it could happen, IMHO

As Ricardo says, CNI is deprecated, so we need to obviously test it. But that can be done with the order of the test modules, what we are already doing. I don't think we need a more complex solution over here. Those are my 2 cents

**Comment by m-dati** [comment]
SLEM PC Aggregates tests [here](https://openqa.suse.de/group_overview/532) `fail` because of incomplete switching management as explained in the PR header. This PR fix it, this is my main effort, but checking that no regression occurrs.

File: lib/containers/podman.pm (hunk 3 above)
Comment (asmorodskyi) [reply lib/containers/podman.pm:44]: in such a case I would rename not only variable but function too

**Comment by mloviska** [comment]
> ...
> 
> > > > and for distros that uses `netavark` as default, we will never switch back to CNI
> > > 
> > > 
> > > Until we have `podma_network_cni` somewhere running, it could happen, IMHO
> > 
> > 
> > As Ricardo says, CNI is deprecated, so we need to obviously test it. But that can be done with the order of the test modules, what we are already doing. I don't think we need a more complex solution over here. Those are my 2 cents
> 
> _SLEM PC Aggregates tests_ [here](https://openqa.suse.de/group_overview/532) `fail` because of incomplete switching management as explained in the PR header. This PR fix it, this is my main effort, but checking that no regression occurrs.
> 
> If we'll CANCEL this PR too those test will remain failing.


According to https://openqa.suse.de/tests/13530849#step/podman_network_cni/20 or https://openqa.suse.de/tests/13530849#step/podman/20 there should not be any switching involved as `netavark` is default over here. Whether this is intentional or not idk

**Comment by m-dati** [comment]
> > and for distros that uses `netavark` as default, we will never switch back to CNI
> 
> Until we have `podma_network_cni` somewhere running, it could happen, IMHO

The fact that netavark is default in such tests, will be correctly managed by the routine I introduced here, by detecting it and avoiding the change, but referring the the actual test structure, we do have a need to switch around version, given the presence of a such `podma_network_cni` test module.

But, if we remove that test, indeed this _improved switch_ need disappears, so question became: _shall we remove that test module_ from here ?

**Comment by m-dati** [comment]
So, agreed to totally re-change the code here and only implement:
Cni exclusion and version control to define netavark default sice SLEM5.5.
Next commit coming soon.

**Review by ricardobranco777 — DISMISSED** [review]

**Comment by m-dati** [comment]
VRs:
- sle-micro-5.1-MicroOS-Image-Updates-x86_64-Build20240214-1-slem_containers_selinux@64bit - 

- sle-micro-5.5-GCE-BYOS-Updates-x86_64-Build20240215-1-
https://openqa.suse.de/tests/13535264

**Comment by m-dati** [comment]
VR

- sle-micro-5.5-GCE-BYOS-Updates-x86_64- sle-micro-5.5-GCE-BYOS-Updates-x86_64-Build20240216-1-slem_containers_selinux.VR27 @64bit
https://openqa.suse.de/tests/13539731

- sle-micro-5.5-GCE-BYOS-Updates-x86_64-Build20240216-1-slem_containers_selinux @64bit - 
https://openqa.suse.de/tests/13539732

- sle-micro-5.2-MicroOS-Image-Updates-s390x-Build20240216-1-slem_containers_selinux @s390x-kvm - 
https://openqa.suse.de/tests/13539733

- sle-micro-5.4-Default-Updates-s390x-Build20240216-1-slem_containers @s390x-kvm - 
https://openqa.suse.de/tests/13539734

**Comment by m-dati** [comment]
VR:

- sle-micro-5.5-GCE-BYOS-Updates-x86_64-Build20240216-1-slem_containers_selinux.VR27 @64bit
  https://openqa.suse.de/tests/13539731

- sle-micro-5.5-GCE-BYOS-Updates-x86_64-Build20240216-1-slem_containers_selinux @64bit
  https://openqa.suse.de/tests/13539732

Regression:

- sle-micro-5.2-MicroOS-Image-Updates-s390x-Build20240216-1-slem_containers_selinux @s390x-kvm 
  https://openqa.suse.de/tests/13539733

- sle-micro-5.4-Default-Updates-s390x-Build20240216-1-slem_containers @s390x-kvm
  https://openqa.suse.de/tests/13539734

- sle-micro-5.1-MicroOS-Image-Updates-aarch64-Build20240218-1-slem_containers_selinux @aarch64
  https://openqa.suse.de/tests/13542367

**Comment by m-dati** [comment]
VR:

- sle-micro-5.5-GCE-BYOS-Updates-x86_64-Build20240216-1-slem_containers_selinux.VR27 @64bit
  https://openqa.suse.de/tests/13539731

- sle-micro-5.5-GCE-BYOS-Updates-x86_64-Build20240216-1-slem_containers_selinux @64bit
  https://openqa.suse.de/tests/13539732

Regression:

- sle-micro-5.2-MicroOS-Image-Updates-s390x-Build20240216-1-slem_containers_selinux @s390x-kvm 
  https://openqa.suse.de/tests/13539733

- sle-micro-5.4-Default-Updates-s390x-Build20240216-1-slem_containers @s390x-kvm
  https://openqa.suse.de/tests/13539734

- sle-micro-5.1-MicroOS-Image-Updates-aarch64-Build20240218-1-slem_containers_selinux @aarch64
  https://openqa.suse.de/tests/13542367

- sle-micro-5.5-Default-Updates-x86_64-Build20240218-1-slem_containers @64bit
  https://openqa.suse.de/tests/13542370

**Comment by m-dati** [comment]
More updated code, needing new review

File: tests/containers/podman_netavark.pm [hunk 4]
```diff
@@ -64,7 +64,7 @@ sub _cleanup {
     validate_script_output('podman network ls', sub { /podman\s+bridge/ });
 }
 
-sub switch_to_netavark {
+sub chk_install_netavark_pkg {
```
Comment (mloviska) [inline tests/containers/podman_netavark.pm:67]: Why?

File: lib/main_containers.pm [hunk 5]
```diff
@@ -107,7 +107,7 @@ sub load_host_tests_podman {
     loadtest 'containers/podman_bci_systemd';
     loadtest 'containers/podman_pods';
     # Default for ALP is Netavark
-    loadtest('containers/podman_network_cni') unless (is_alp || is_sle_micro('6.0+'));
+    loadtest('containers/podman_network_cni') unless (is_alp || is_sle_micro('5.5+'));
```
Comment (mloviska) [inline lib/main_containers.pm:110]: I thought slem 5.5 is a specific case only for PC as the image is newer than non PC slem 5.5

**Review by mloviska — CHANGES_REQUESTED** [review]

File: lib/main_containers.pm (hunk 5 above)
Comment (m-dati) [reply lib/main_containers.pm:110]: Yes, probably I should add a check around `...is_publiccloud` too, but in what case?
_exclude_ cni for something like (is_pubc and is_slemicro=5.5) ?

Comment (mloviska) [reply lib/main_containers.pm:110]: yes

File: tests/containers/podman_netavark.pm (hunk 4 above)
Comment (m-dati) [reply tests/containers/podman_netavark.pm:67]: The _original_ master version of switch routine had 2 phases:
n.1 install netavark and aadvark checking installation type transactional or not, 
n.2 restart system to make netavark new default.

But that  switch routine was called only when `is_cni` therefore not performed in a case, new added slem 5.5 skipped the is_cni test, but needing transactional aadvark, so that it failed in https://openqa.suse.de/tests/13535263#step/podman_netavark/120.

The `...vark` packages in this module should be _always_ present-or-installed-if-missing; then in next phase activated if is_cni was true, therfere the `....vark` install phase should be always executed(without package pre-selection, being automatically skipped if already present) in a **transactional check test**, to define the _install type_.

This last is performed in the new routine `chk_install_netavark_pkg` (that can be eventually renamed, if needed) splitted by old `switch_to_..`.  
The new `switch_to_..` routine contains now the only system restart phase.

File: tests/containers/podman_netavark.pm [hunk 6]
```diff
@@ -93,11 +95,11 @@ sub run {
         return 1;
     }
 
+    chk_install_netavark_pkg;
     if (is_cni_default || is_cni_in_tw) {
-        switch_to_netavark;
+        switch_to_netawark;
```
Comment (m-dati) [inline tests/containers/podman_netavark.pm:100]: I used an incorrect name `netawark`, but not affecting the code syntax. I will update it in next commits.

File: tests/containers/podman_netavark.pm [hunk 7]
```diff
@@ -24,7 +24,7 @@ sub is_cni_in_tw {
 # podman >=4.8.0 defaults to netavark
 sub is_cni_default {
     my $podman_version = get_podman_version();
-    return package_version_cmp($podman_version, '4.8.0') < 0 || is_sle_micro('<6.0');
+    return package_version_cmp($podman_version, '4.8.0') < 0 || is_sle_micro('<5.5');
```
Comment (m-dati) [inline tests/containers/podman_netavark.pm:27]: IMHO in this check is missing the same `...NetworkBackend...=~ "cni"`, that is present in `is_cni_in_tw` test.
In this actual status the check doesn't consider the current default.
I wish to update it accordingly

**Comment by m-dati** [comment]
tests after last commit:

VR:
- sle-micro-5.5-GCE-BYOS-Updates-x86_64-Build20240216-1-slem_containers_selinux
  https://openqa.suse.de/tests/13545172

Regression:
- sle-micro-5.3-MicroOS-Image-Updates-x86_64-Build20240218-1-slem_containers_selinux
  https://openqa.suse.de/tests/13545584
- sle-micro-5.4-Default-Updates-s390x-Build20240216-1-slem_containers
  https://openqa.suse.de/tests/13545312
- sle-micro-5.1-MicroOS-Image-Updates-aarch64-Build20240218-1-slem_containers_selinux
  https://openqa.suse.de/tests/13545310
- sle-micro-5.5-Default-Updates-x86_64-Build20240218-1-slem_containers
  https://openqa.suse.de/tests/13545153

File: lib/main_containers.pm (hunk 5 above)
Comment (m-dati) [reply lib/main_containers.pm]: done

File: tests/containers/podman_netavark.pm [hunk 8]
```diff
@@ -64,16 +64,20 @@ sub _cleanup {
     validate_script_output('podman network ls', sub { /podman\s+bridge/ });
 }
 
-sub switch_to_netavark {
-    my @pkgs = qw(netavark aardvark-dns);
+sub install_pkg {
```
Comment (ricardobranco777) [inline tests/containers/podman_netavark.pm:67]: Please make a separate commit for this change.

**Comment by mloviska** [comment]
Please run more VRs with TW and MicroOS, especially the distro update test suites as `container-host-old2microosnext` or `container-host2microosnext`

File: tests/containers/podman_netavark.pm (hunk 8 above)
Comment (m-dati) [reply tests/containers/podman_netavark.pm]: done

**Comment by m-dati** [comment]
> Please run more VRs with TW and MicroOS, especially the distro update test suites as `container-host-old2microosnext` or `container-host2microosnext`

Please, kind help to find the links to the desired test groups.

**Review by ricardobranco777 — APPROVED** [review]

**Comment by m-dati** [comment]
Finally here are the VRs on OOO, PASS OK:

cloned from [OOO](https://openqa.opensuse.org/tests/overview?distri=microos&distri=opensuse&version=Tumbleweed&build=20240221&groupid=1):

- opensuse-Tumbleweed-JeOS-for-kvm-and-xen-x86_64-Build20240221-jeos-container_host @64bit_virtio-2G
 [3956790](https://openqa.opensuse.org/tests/3956790)

- microos-Tumbleweed-MicroOS-Image-ContainerHost-x86_64-Build20240221-container-host 
  [3956780](https://openqa.opensuse.org/tests/3956780)

- microos-Tumbleweed-MicroOS-Image-ContainerHost-x86_64-Build20240221-container-host2microosnext 
  [3956781](https://openqa.opensuse.org/tests/3956781)

- microos-Tumbleweed-MicroOS-Image-ContainerHost-x86_64-Build20240221-container-host-old2microosnext
  [3956803](https://openqa.opensuse.org/tests/3956803)
  1) cloned-from 3955651 failing(always) in [rootless_podman/51] (https://openqa.opensuse.org/tests/3955651#step/rootless_podman/51);
  2) the cloned test passed ok the modules updated in this PR, then failed in same way as (1).

**Comment by m-dati** [comment]
code also rebased, in last push

**Comment by m-dati** [comment]
@mloviska @asmorodskyi please, kind look

**Review by mloviska — APPROVED** [review]

**Comment by DimStar77** [comment]
https://openqa.opensuse.org/tests/3960079#investigation

Test sens broken by this change
